rasa/shared/utils/common.py

```diff
@@ -1,35 +1,44 @@
 import asyncio
 import functools
 import importlib
+import inspect
 import logging
 from typing import Text, Dict, Optional, Any, List, Callable, Collection
 
 logger = logging.getLogger(__name__)
 
 
 def class_from_module_path(
-    module_path: Text, lookup_path: Optional[Text] = None
+    module_path: Text, lookup_path: Optional[Text] = None, ensure_class: bool = False
 ) -> Any:
     """Given the module name and path of a class, tries to retrieve the class.
 
     The loaded class can be used to instantiate new objects."""
     # load the module, will raise ImportError if module cannot be loaded
+    klass = None
     if "." in module_path:
         module_name, _, class_name = module_path.rpartition(".")
         m = importlib.import_module(module_name)
         # get the class, will raise AttributeError if class cannot be found
-        return getattr(m, class_name)
+        klass = getattr(m, class_name)
     else:
-        module = globals().get(module_path, locals().get(module_path))
+        klass = globals().get(module_path, locals().get(module_path))
         if module is not None:
-            return module
+            klass = module
 
-        if lookup_path:
+        if klass is None and lookup_path:
             # last resort: try to import the class from the lookup path
             m = importlib.import_module(lookup_path)
-            return getattr(m, module_path)
-        else:
-            raise ImportError(f"Cannot retrieve class from path {module_path}.")
+            klass = getattr(m, module_path)
+
+    if klass is None:
+        raise ImportError(f"Cannot retrieve class from path {module_path}.")
+
+    if ensure_class and not inspect.isclass(klass):
```

Comment:
I don't know enough of the usage of this function, but the name `class_from_module_path ()` suggests that it should always be returning a class, whereas the code at line 25 suggests otherwise. Would need feedback on this

if it's expected to return a class all the time, i'd make `ensure_class` the default (effectively removing it from the func signature), and raise a deprecation warning.

(the code I wrote for this one is pretty basic, mainly because I wanted to show a draft version earlier)

Reply:
I do think it needs to always return a class.

The condition in L19 is for the two different cases `mypackage.MyModel` and `MyModel` as they need to be loaded in different ways. In the case of the later (handled in the `else` part) the module is already in scope and hence needs to be retrieved from there (this is the `globals` and `locals` part) - this should have gotten a link to a stackoverflow post where this likely has its origin 😅